### PR TITLE
Add an `approveWithRevoke` function to atomically update permissions

### PR DESCRIPTION
**Pull request by amiecorso**
BW-618

Adds an `approveWithRevoke` function to atomically revoke and approve an old and new spend permission, respectively.  Also receives a `PeriodSpend` struct, which will be used to indicate the last updated period at the time the replace request was made. When called, if there has been ANY subsequent spending behavior against the old spend permission, indicated by differences in the submitted vs. current value of this PeriodSpend struct, the transaction reverts. This is intended to prevent frontrunning a commitment to update a permission with extractive spending on the old permission.

Adds a `getLastUpdatedPeriod` public function to get access not to the current period, but to the latest stored cumulative spend activity for the old permission.

The only field that must be shared between the old and new permissions is the `account` that is authorizing the transition.  Everything else can differ and there is not state carryover between the old and new permissions' spend accounting.

### CODE REVIEW

File: src/SpendPermissionManager.sol [hunk 1]
```diff
@@ -199,6 +208,20 @@ contract SpendPermissionManager is EIP712 {
         _;
     }
 
+    /// @notice Require that two spend permissions have the same `account` field.
+    ///
+    /// @param firstSpendPermission First spend permission to compare.
+    /// @param secondSpendPermission Second spend permission to compare.
+    modifier requireSameAccount(
```
Comment (ilikesymmetry) [inline src/SpendPermissionManager.sol:215]: seems like this modifier is only going to be used once, so would prefer to have this check in the function itself for easier reading

File: src/SpendPermissionManager.sol [hunk 2]
```diff
@@ -279,15 +302,44 @@ contract SpendPermissionManager is EIP712 {
         _transferFrom(spendPermission.token, spendPermission.account, spendPermission.spender, value);
     }
 
+    /// @notice Revokes a spend permission and approves a new one atomically.
```
Comment (ilikesymmetry) [inline src/SpendPermissionManager.sol:305]:
```suggestion
    /// @notice Approves a permission while revoking another if it's last update has not changed.
```

File: src/SpendPermissionManager.sol [hunk 3]
```diff
@@ -279,15 +302,44 @@ contract SpendPermissionManager is EIP712 {
...
+    ///
+    /// @dev Enforces that the last updated period of the permission being revoked matches the last valid updated period
+    ///      submitted as an argument. This is to prevent frontrunning `replace` with additional last-minute spends.
+    ///
+    /// @dev The `account` of the permissions must match, but the remaining fields can differ.
+    ///
+    /// @dev Can only be called by the `account` of a permission.
```
Comment (ilikesymmetry) [inline src/SpendPermissionManager.sol:312]: nit: remove extra line spaces between shared `@dev` tags

File: src/SpendPermissionManager.sol [hunk 4]
```diff
@@ -279,15 +302,44 @@ contract SpendPermissionManager is EIP712 {
...
+    ///
+    /// @param permissionToRevoke Details of the spend permission to revoke.
+    /// @param permissionToApprove Details of the spend permission to approve.
+    /// @param lastValidUpdatedPeriod Last valid updated period for the spend permission being revoked.
+    function replace(
+        SpendPermission calldata permissionToRevoke,
+        SpendPermission calldata permissionToApprove,
+        PeriodSpend calldata lastValidUpdatedPeriod
```
Comment (ilikesymmetry) [inline src/SpendPermissionManager.sol:320]: What do you think of this renaming?
```suggestion
    function approveWithRevoke(
        SpendPermission calldata permissionToApprove,
        SpendPermission calldata permissionToRevoke,
        PeriodSpend calldata requiredLastUpdatedPeriod
```

File: src/SpendPermissionManager.sol [hunk 5]
```diff
@@ -279,15 +302,44 @@ contract SpendPermissionManager is EIP712 {
...
+        if (
+            lastUpdatedPeriod.spend != lastValidUpdatedPeriod.spend
+                || lastUpdatedPeriod.start != lastValidUpdatedPeriod.start
+                || lastUpdatedPeriod.end != lastValidUpdatedPeriod.end
+        ) {
+            revert InvalidLastUpdatedPeriod(lastUpdatedPeriod, lastValidUpdatedPeriod);
+        }
+        // revoke old and approve new spend permissions
+        _revoke(permissionToRevoke);
+        _approve(permissionToApprove);
```
Comment (ilikesymmetry) [inline src/SpendPermissionManager.sol:333]: do we want to enforce that these two are not the same permission?

File: test/src/SpendPermissions/replace.t.sol [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+            salt: salt,
+            extraData: extraData
+        });
+        vm.startPrank(sender);
+        vm.expectRevert(abi.encodeWithSelector(SpendPermissionManager.InvalidSender.selector, sender, account));
+        mockSpendPermissionManager.replace(existingSpendPermission, newSpendPermission, lastValidUpdatedPeriod);
+        vm.stopPrank();
+    }
+
+    function test_replace_revert_invalidLastUpdatedPeriod(
```
Comment (ilikesymmetry) [inline test/src/SpendPermissions/replace.t.sol:53]: Can we check more variants of this explicitly? For example, it should fail both if we're in the same period with incremented spend or same spend in a different period boundary?

File: src/SpendPermissionManager.sol (hunk 4 above)
Comment (amiecorso) [reply src/SpendPermissionManager.sol:323]: i like the emphasis on the approval that's happening

File: src/SpendPermissionManager.sol (hunk 5 above)
Comment (amiecorso) [reply src/SpendPermissionManager.sol:336]: personally I don't... we don't enforce that batches don't contain duplicates, nor do we enforce that permissions being approved aren't already approved (or permissions being revoked aren't already revoked), and in general have taken the approach of avoiding validation unless there's a risk of inconsistent state onchain. adding validation here would feel arbitrary.

File: test/src/SpendPermissions/approveWithRevoke.t.sol [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+                lastValidUpdatedPeriod
+            )
+        );
+        mockSpendPermissionManager.approveWithRevoke(
+            newSpendPermission, existingSpendPermission, lastValidUpdatedPeriod
+        );
+        vm.stopPrank();
+    }
+
+    function test_approveWithRevoke_revert_mismatchedAccounts(
```
Comment (ilikesymmetry) [inline test/src/SpendPermissions/approveWithRevoke.t.sol:213]: nit: usually like to put all the revert cases up top in the order they appear in the code so this should come immediately after the invalid sender check

**Review by ilikesymmetry — APPROVED** [review]
great tests, nice work!

**Pull request merged**